doc/manual/man/man1/makeSCF.1
.TH "makeSCF" 1 "" "" "Staden Package"
.SH "NAME"
.PP
makeSCF \- Converts trace files to SCF files.

.SH "SYNOPSIS"
.PP

\fBmakeSCF\fP [\fB-8\fP] [\fB-2\fP] [\fB-3\fP]
-(\fBabi\fP|\fBalf\fP|\fBscf\fP|\fBpln\fP) \fIinput_name\fP
[\fB-compress\fP \fIcompression_mode\fP] [\fB-normalise\fP
[\fB-output\fP \fIoutput_name\fP]

.SH "DESCRIPTION"
.PP

\fBMakeSCF\fP converts trace files to the SCF format. It can input ABI 373A,
Pharmacia A.L.F., or previously created SCF files (although converting from
SCF to SCF serves no useful purpose!). 

.SH "OPTIONS"
.PP

.TP
\fB-8\fP
Force conversion to 8 bit sample data. This shrinks the size of SCF
files using 16 bit sample values, but at a loss of resolution. For trace
display purposes this accuracy loss is acceptable.

.TP
\fB-2\fP
Force the output to be written in SCF version 2. By default the
latest version (3) is used.

.TP
\fB-3\fP
Force the output to be written in SCF version 3. This is the default.

.TP
\fB-s\fP
Silent mode. This prevents the output of the copyright message.

.TP
\fB-abi\fP, \fB-alf\fP, \fB-scf\fP, \fB-any\fP
Specify an input file format. A file format of "any" will force
\fBmakeSCF\fP to automatically determine the correct input file type.

.TP
\fB-compress\fP \fIcompression_mode\fP
Requests the generated SCF file to be passed through a separate compression
program before writing to disk. \fBmakeSCF\fP does not contain any
compression algorithms itself. It requires the appropriately named tool to
be on the system and in the user's @r{PATH}.
Valid responses for \fIcompression_mode\fP are (in order of best compression
first) \fBbzip\fP, \fBgzip\fP, \fBcompress\fP and \fBpack\fP. Note
that \fBbzip\fP at present is only bzip version 1 and that bzip version 2
is incompatible.

.TP
\fB-normalise\fP
Performs some very simple trace normalisation. This subtracts the
background signal (by defining the background signal to be the lowest of
the four traces) and rescales the peak heights, averaging the height over
a `window' of 1000 trace sample points. This option may be useful
for some unscaled ALF files.

.TP
\fB-output\fP \fIfile\fP
Specifies the filename for the SCF file to be produced. If this is not
specified the SCF file will be sent to standard output.
.TE
.SH "EXAMPLES"
.PP

To convert an ABI 373A trace:

.nf
.in +0.5i
\fBmakeSCF -8 -abi trace.abi -output trace.scf\fP
.in -0.5i
.fi

To convert an ALF archive to individual SCF files (Warning! this 
will most certainly fail if your clone names contain spaces):

.nf
.in +0.5i
\fBalfsplit trace.alf | awk '/^Clone/ {print $3 "ALF"@\fP' > trace.files}

\fBsh -c 'for i in `cat trace.files`;do makeSCF -alf $i -output\fP
\fB    $i.scf;done\fP
.in -0.5i
.fi

.SH "NOTES"
.PP

If ABI and A.L.F files are edited before input to makeSCF the contents of
the resulting SCF files are unpredictable.
To use Pharmacia A.L.F. files the \fBalfsplit\fP program should first
be used. Then \fBmakeSCF\fP should be run on each of the split files.
See the example above.

.SH "SEE ALSO"
.PP

_fxref(Formats-Scf, scf(4), formats)
_fxref(Man-convert_trace, convert_trace(1), convert_trace.1)
_fxref(Man-eba, eba(1), eba.1)



NEWS
-----------------------------------------------------------------------------
30th September 2005
===================

Version 1.6.0 tagged

-----------------------------------------------------------------------------

27th September 2005
===================

Gap4
----

* Renamed the discrepancy plot to be second highest confidence plot.

* Fixed some (apparently rare) find internal joins crashes.

* Added a minimum quality slider to the Report Mutations "by
  differences" mode. Also fixed crashes and missing pictures in the
  HTML report.

io_lib
------

* Changed the HashFile format and updated it to be official version
  1.00.

  The key difference is that there's now a footer pointing back into
  the hashfile header allowing for the index to simply be appended to
  a (e.g. tar) archive.

* Improved handling of 454 data in ZTR; it no longer attempts to
  compress raw pyrosequencing data as this was not working well.

* Fixed Bug ID 1289095 (file descriptor leak).

-----------------------------------------------------------------------------

24th August 2005
================

Gap4
----

* Rewrote the Shuffle Pads algorithm to be more like ReAligner. It can
  introduce new pads and remove old ones as necessary in order to
  optimise the alignments. It has also moved from inside the contig
  editor to part of the main menu so that it can be scripted easier.

* The join editor invoked from Find Read Pairs plot now brings up the
  editors at the start/end of the sequences forming the read-pair
  instead of the start/end of the two contigs.

* Search by sequence in the editor now finds consensus matches
  too. Also fixed some bugs where cutoff data matches were not always
  found and in some cases (reverse search) non-cutoff matches could be
  missed.

* The SNP Candidates plot now allows for the consensus to be saved
  directly from this window. The consensus and fake sequences added
  when a real split is requested now includes the "unsplit" readings
  from regions which apparently contained no SNPs (and therefore
  belong to all split contigs).

* Report Mutations bug fix. Producing HTML output now works (albeit
  with less content) when asking for SNPs to be derived from
  "differences" instead of tags.

Trev
----

* Pyrosequencing data (eg 454 flowgrams) are now supported. Trev
  auto-senses the type and views accordingly. Added format for 454's
  SFF standard too.

Io_lib
------

* New release - 1.9.0. No longer uses libtool in configuration.

* Added "HASH=hashfile" component of RAWDATA search path along with
  hash_tar and hash_extract tools. These operate like index_tar but
  provide true fast random access into tar files (or other formats if
  needed).

* Major update of I/O mechanisms to use in-memory equivalents of the
  files being read from or written to. This has lead to some
  incompatibilities (see src/io_lib/CHANGES), but is also an order of
  magnitude faster in many cases.

Spin
----

* Under MS Windows spin and EMBOSSwin should coexist happily to the
  extent that Spin will spot the EMBOSSwin installation and use it.

Pregap4
-------

* Fixed various space-in-pathname bugs (most prevalent on MS Windows).


Misc
----

* Removed some redundant old ABI utilities: getABIfloat, getABIhex,
  getABIraw, getABIstring. All of these can be achieved via getABIfield.


-----------------------------------------------------------------------------

3 May 2005
==========

Internal Sanger test version rel-1-6-0b3 ("gap4 -test")

Gap4
----

* Substantial modifications to the gap4 I/O mechanisms (balancing the
  "freetree"). The upshot of this is that the worst-case performance
  is now dramatically improved and the average performance will be
  slightly better too. Complementing a 10Mb contig dropped from about
  50 minutes to 1.5 minutes. Disassembling all readings (approx
  175000) in this database went from ~3 hours to 2.5 minutes.

* The contig editor "sequence sets" are now coloured to aid
  distinguishing.

* Fixed bugs with X11 BadDrawable crashes (I hope).

* Reverted the OLIGo tag back to using "serial=#" as the first line,
  although the additional information is still kept (albeit in a
  reduced precision format to make the tag shorter).

-----------------------------------------------------------------------------

29 March 2005
=============

Internal Sanger test version rel-1-6-0b2 (the previous "gap4 -test")

Gap4
----

* Added 64-bit filesize support, meaning that Gap4 databases can now
  be larger than 2Gb in file size. Specify "-bitsize 64" to force
  64-bit mode, otherwise databases will still be created in the old
  32-bit format. Copy_db has a -b flag to change bitsize when copying
  databases so this aids in migration.

* A new experimental option "SNP candidates" (in the main view
  menu). Primarily designed for pulling apart the two alleles from
  mixed shotguns, it also has uses for exploring and sometimes
  resolving collapsed repeats. User feedback at this stage is most
  welcome.

* The consistency plots now distinguish between the old "discrepancy
  plot" (now named diploid discrepancy plot) and a new discrepancy
  plot that takes does not have the "two alleles only in 50/50 ratios"
  detector. This makes it more suitable for finding potential
  collapsed repeats.

* The contig editor now has the notion of "sets". A set is basically a
  sub-group of a contig, which is useful for exploring the outcome
  within a conti of pulling apart a collapsed repeat without actually
  doing it. The editor is automatically invoked in this manner by the
  SNP candidates output. This is largely experimental, but there are a
  number of potential ways to improve interaction. Ideas on a postcard
  please!

* Lots of speedups (primarily for the huge 64-bit databases):
  - Drawing of some consistency graphs;
  - Experiment file reading on large files (1000 fold on a 1Mb sequence);
  - Dealing with lists of contigs (eg "all contigs") is now O(N)

src/io_lib/progs/makeSCF.c
/*
 * Copyright (c) Medical Research Council 1994-1998. All rights reserved.
 *
 * Permission to use, copy, modify and distribute this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * this copyright and notice appears in all copies.
 *
 * This file was written by James Bonfield, Simon Dear, Rodger Staden,
 * as part of the Staden Package at the MRC Laboratory of Molecular
 * Biology, Hills Road, Cambridge, CB2 2QH, United Kingdom.
 *
 * MRC disclaims all warranties with regard to this software.
 */


/*
 * makeSCF v3.06, 17/04/2001
 *
 * Derived from the older makeSCF; this one has been rewritten to use the new
 * IO libraries and no longer performs quality clipping itself. It also writes
 * in a new format that is more easily compressed.
 */

#include <stdio.h>
#include "Read.h"
#include "traceType.h"
#include "xalloc.h"

/*
 * Add our comments.
 * 1. Work out the comment length - simply add 1K to the current length.
 * 2. Copy the old comments to our new.
 * 3. Replace '\n' with ' '.
 * 4. Add our comments and switch.
 */
void add_comments(Read *r, char *name, int format) {
    Comments *cc;
    int clen;
    char *cp;
    char buf[1024];

    /* 1. */
    if (r->info) {
	clen = strlen(r->info) + 1024;
    } else {
	clen = 1024;
    }

    /* 2. */
    if (NULL == (cc = (char *)xmalloc(clen)))
	return;

    if (r->info)
	strcpy(cc, r->info);
    else
	*cc = 0;
    
    /* 3. */
    cp = cc;
/*
    while (*cp) {
	if (*cp == '\n')
	    *cp = ' ';
	
	cp++;
    }
    *cp++ = '\n';
*/
  
    /* 4. */
    sprintf(buf, "CONV=makeSCF V3.06\nDATF=%s\nDATN=%s\n",
	    trace_type_int2str(format), name);

    strcat(cp, buf);

    if (r->info)
	xfree(r->info);

    r->info = cc;
}

void scale_trace8(Read *r) {
    double s;
    int i;

    if (r->maxTraceVal <= 255)
	return;
    
    s = ((double)255)/r->maxTraceVal;

    for (i = 0; i < r->NPoints; i++) {
	r->traceA[i] *= s;
	r->traceC[i] *= s;
	r->traceG[i] *= s;
	r->traceT[i] *= s;
    }

    r->maxTraceVal = 255;
}

/*
 * Here we just take the minimum trace value and subtract this from all others.
 * The assumption is that the signal will always be 'base line' on at least
 * one of the four channels.
 */
void subtract_background(Read *r) {
    int i, min;
    for (i = 0; i < r->NPoints; i++) {
	min = 999999;
	if (r->traceA[i] < min) min = r->traceA[i];
	if (r->traceC[i] < min) min = r->traceC[i];
	if (r->traceG[i] < min) min = r->traceG[i];
	if (r->traceT[i] < min) min = r->traceT[i];
	r->traceA[i] -= min;
	r->traceC[i] -= min;
	r->traceG[i] -= min;
	r->traceT[i] -= min;
    }
}

/*
 * Find the average background level of a trace, and subtract this from the
 * peak heights.
 *
 * NB. That method is flawed. For now take the minimum instead of average, but
 * this also has horrid flaws. See above method.
 */
void subtract_background_old(Read *r) {
    int i, j, min, bg, max = 0;
    int win_len = 501, win_len2 = win_len/2;
    int *background;

    if (NULL == (background = (int *)xmalloc((r->NPoints + 2 * win_len)
					     * sizeof(*background))))
	return;

    if (r->NPoints < win_len)
	win_len = r->NPoints;

    /* Find minimum trace levels at each point */
    for (i = 0, j = win_len2; i < r->NPoints; i++, j++) {
	min = INT_MAX;
	if (r->traceA[i] < min)
	    min = r->traceA[i];
	if (r->traceC[i] < min)
	    min = r->traceC[i];
	if (r->traceG[i] < min)
	    min = r->traceG[i];
	if (r->traceT[i] < min)
	    min = r->traceT[i];
	background[j] = min;
    }
    for (i = 0; i < win_len2; i++) {
	background[i] = background[i + win_len2];
	background[i + r->NPoints] = background[i + r->NPoints - win_len2];
    }

    /* Take lowest background over win_len and subtract it */
    for (i = 0; i < r->NPoints; i++) {
	/* Could optimise this considerably */
	bg = INT_MAX;
	for (j = 0; j < win_len; j++) {
	    if (background[i + j] < bg)
		bg = background[i + j];
	}

	r->traceA[i] -= bg;
	r->traceC[i] -= bg;
	r->traceG[i] -= bg;
	r->traceT[i] -= bg;

	if (r->traceA[i] > max) max = r->traceA[i];
	if (r->traceC[i] > max) max = r->traceC[i];
	if (r->traceG[i] > max) max = r->traceG[i];
	if (r->traceT[i] > max) max = r->traceT[i];
    }
    
    r->maxTraceVal = max;

    xfree(background);
}

/*
 * Find the maximum height of traces at the called bases. Use this to clip any
 * other bases.
 */
void reset_max_called_height(Read *r) {
    int i, max = 0;

    /* Find max */
    for (i=0; i < r->NBases; i++) {
	switch(r->base[i]) {
	case 'a':
	case 'A':
	    if (r->traceA[r->basePos[i]] > max)
		max = r->traceA[r->basePos[i]];
	    break;

	case 'c':
	case 'C':
	    if (r->traceC[r->basePos[i]] > max)
		max = r->traceC[r->basePos[i]];
	    break;

	case 'g':
	case 'G':
	    if (r->traceG[r->basePos[i]] > max)
		max = r->traceG[r->basePos[i]];
	    break;

	case 't':
	case 'T':
	    if (r->traceT[r->basePos[i]] > max)
		max = r->traceT[r->basePos[i]];
	    break;
	}
    }

    /* Clip to max */
    for (i = 0; i < r->NPoints; i++) {
	if (r->traceA[i] > max)
	    r->traceA[i] = max;
	if (r->traceC[i] > max)
	    r->traceC[i] = max;
	if (r->traceG[i] > max)
	    r->traceG[i] = max;
	if (r->traceT[i] > max)
	    r->traceT[i] = max;
    }
    if (r->maxTraceVal > max)
	r->maxTraceVal = max;
}

void rescale_heights(Read *r) {
    int win_len = 1000;
    int total = 0;
    int max, max2;
    int i, j, k;
    double max_val = 0, rescale;
    TRACE *ta, *tc, *tg, *tt;

    ta = r->traceA;
    tc = r->traceC;
    tg = r->traceG;
    tt = r->traceT;

    if (r->NPoints < 2*win_len + 1)
	return;

    for (k = 0; k < 2; k++) {
	max2 = win_len * r->maxTraceVal;
	
	for (i = 0; i < win_len; i++) {
	    max = 0;
	    if (ta[i] > max) max = ta[i];
	    if (tc[i] > max) max = tc[i];
	    if (tg[i] > max) max = tg[i];
	    if (tt[i] > max) max = tt[i];
	    total += max;
	}
	
	for (j = 0; i < r->NPoints; i++, j++) {
	    max = 0;
	    if (ta[j] > max) max = ta[j];
	    if (tc[j] > max) max = tc[j];
	    if (tg[j] > max) max = tg[j];
	    if (tt[j] > max) max = tt[j];
	    total -= max;
	    
	    max = 0;
	    if (ta[i] > max) max = ta[i];
	    if (tc[i] > max) max = tc[i];
	    if (tg[i] > max) max = tg[i];
	    if (tt[i] > max) max = tt[i];
	    total += max;
	    
	    if (k == 0) {
		if (r->traceA[j] * ((double)max2 / total) > max_val)
		    max_val = r->traceA[j] * ((double)max2 / total);
		if (r->traceC[j] * ((double)max2 / total) > max_val)
		    max_val = r->traceC[j] * ((double)max2 / total);
		if (r->traceG[j] * ((double)max2 / total) > max_val)
		    max_val = r->traceG[j] * ((double)max2 / total);
		if (r->traceT[j] * ((double)max2 / total) > max_val)
		    max_val = r->traceT[j] * ((double)max2 / total);
	    } else {
		r->traceA[j] *= (double)max2 / total * rescale;
		r->traceC[j] *= (double)max2 / total * rescale;
		r->traceG[j] *= (double)max2 / total * rescale;
		r->traceT[j] *= (double)max2 / total * rescale;
	    }
	}

	for (; j < r->NPoints; j++) {
	    if (k == 0) {
		if (r->traceA[j] * ((double)max2 / total) > max_val)
		    max_val = r->traceA[j] * ((double)max2 / total);
		if (r->traceC[j] * ((double)max2 / total) > max_val)
		    max_val = r->traceC[j] * ((double)max2 / total);
		if (r->traceG[j] * ((double)max2 / total) > max_val)
		    max_val = r->traceG[j] * ((double)max2 / total);
		if (r->traceT[j] * ((double)max2 / total) > max_val)
		    max_val = r->traceT[j] * ((double)max2 / total);
	    } else {
		r->traceA[j] *= (double)max2 / total;
		r->traceC[j] *= (double)max2 / total;
		r->traceG[j] *= (double)max2 / total;
		r->traceT[j] *= (double)max2 / total;
	    }
	}

	if (max_val > 65535)
	    rescale = 65535 / max_val;
	else
	    rescale = 1.0;
    }
}

static int convert(char *in, mFILE *ofp, char *out, int format, int prec,
		   int comp, int normalise) {
    Read *r;

    if (NULL == (r = read_reading(in, format))) {
	fprintf(stderr, "%s: failed to read\n", in);
	return 1;
    }

    if (normalise) {
	subtract_background(r);
	reset_max_called_height(r);
	rescale_heights(r);
    }

    add_comments(r, in, format);
    if (prec == 1)
	scale_trace8(r);

    if (comp != -1)
	set_compression_method(comp);
    if (0 != (mfwrite_reading(ofp, r, TT_SCF))) {
	fprintf(stderr, "%s: failed to write\n", out);
	read_deallocate(r);
	return 1;
    }

    read_deallocate(r);
    return 0;
}


void usage(void) {
    fprintf(stderr,
	    "makeSCF [-8] [-2] [-3] [-s] [-compress mode] [-normalise]\n"
	    "       -(abi|alf|scf|any) input_name [-output output_name]\n"
	    " or\n"
	    "makeSCF [-8] [-2] [-3] [-s] [-compress mode] [-normalise]\n"
	    "       [-(abi|alf|scf|any)] input_name1 output_name1 ... "
	    "input_nameN output_nameN \n");

    exit(1);
}

int main(int argc, char **argv) {
    int format = TT_ANY, r, prec = 0, version = 3, silent = 0;
    int compress_mode = -1;
    char *inf = NULL;
    char *outf = NULL;
    mFILE *ofp = mstdout();
    int normalise = 0;

    for (argc--, argv++; argc > 0; argc--, argv++) {
	if (strcmp(*argv, "-8") == 0) {
	    prec = 1;
	} else if (strcmp(*argv, "-2") == 0) {
	    version = 2;
	} else if (strcmp(*argv, "-3") == 0) {
	    version = 3;
	} else if (strcmp(*argv, "-normalise") == 0) {
	    normalise = 1;
	} else if (strcmp(*argv, "-s") == 0) {
	    silent = 1;
	} else if (strcasecmp(*argv, "-abi") == 0) {
	    format = TT_ABI;
	    inf = *++argv;
	    argc--;
	} else if (strcasecmp(*argv, "-alf") == 0) {
	    format = TT_ALF;
	    inf = *++argv;
	    argc--;
	} else if (strcasecmp(*argv, "-scf") == 0) {
	    format = TT_SCF;
	    inf = *++argv;
	    argc--;
	} else if (strcasecmp(*argv, "-any") == 0) {
	    format = TT_ANY;
	    inf = *++argv;
	    argc--;
	} else if (strcasecmp(*argv, "-output") == 0) {
	    outf = *++argv;
	    argc--;
	} else if (strcasecmp(*argv, "-compress") == 0) {
	    compress_mode = compress_str2int(*++argv);
	    argc--;
	} else {
	    break;
	}
    }

    /* if no args left than input file must have been specified */
    if (!argc && !inf)
	usage();

    /* if outfile set, then using original syntax, so don't expect
       any extra args */
    if (argc && outf)
      usage();

    if (!silent) {
	printf("makeSCF v3.06\n");
	printf("Copyright (c) MRC Laboratory of Molecular Biology, 2001. All rights reserved.\n");
    }

    set_scf_version(version);


    if(!argc) {
	/* original calling syntax */
	if (outf) {
	    ofp = mfopen(outf, "wb+");
	    if (NULL == ofp) {
		perror(outf);
		return 1;
	    }
	}

	r = convert(inf, ofp, outf, format, prec, compress_mode, normalise);
	mfclose(ofp);

	return r;

    }

    /* else */ {
	/* new calling syntax, handling multiple files */
	int result=0;

	for (; argc > 0; argc--, argv++) {
	    if (inf) {
		/* got infile, so get outfile and process */
		outf= *argv;
		ofp = mfopen(outf, "wb+");
		
		if (NULL == ofp) {
		    perror(outf);
		    if(!result) result=1;
		    continue;
		}
		r = convert(inf, ofp, outf, format, prec, compress_mode,
			    normalise);
		mfclose(ofp);
		if(!result) /* keep track of the first error */
		    result=r;
	      
		/* now need to get another infile */
		inf=NULL;
	    } else {
		/* need infile */
		inf= *argv;
	    }
	}

	return result;
    }
}

